### ISSUE TITLE
Fixed cart discount calculated incorrectly when product first added to cart.

### ISSUE THREAD

**Issue by aligent-lturner** [magento/magento2#26622]
### Preconditions (*)

1. Magento 2.3.4
2. PHP 7.1+
3. Create a cart price rule for a fixed amount for the whole cart (e.g. $1). The rule should have no conditions attached to it (or restrictions on the items it applies to)
4. Create at least one configurable product with at least one simple product attached to each configurable.
5. Set the price of each item to be $10 (actual price doesn't matter - just used $10 for example)

### Steps to reproduce (*)

1. Navigate to PDP of configurable product
2. Select options for the configurable and add to cart

### Expected result (*)

1. Item is added to cart
2. Cart total is $9 ($10 for item - $1 discount)

### Actual result (*)

1. Item is added to cart
2. Cart total is $9.50 ($10 for item - $0.50 discount)

### Notes
The issue is also present any time a new simple product is added to the cart. For example, if a different configurable option was chosen, the discount would be calculated as $0.67 (then $0.75 for another). The reason for this behaviour is due to the following:

When a configurable is added to the cart, two items actually get added as Quote items - the configurable, and the simple item representing the selected options.
As part of the rule calculation, it skips child (simple) items by checking the result of `getParentItemId()`:

https://github.com/magento/magento2/blob/e6e6725329f75e36380e788e254ad012e8ae0759/app/code/Magento/SalesRule/Model/Validator.php#L385-L388

However, for a newly added item, `parentItemId` is not set yet, as this is only set in the `beforeSave` function of the quote item:

https://github.com/magento/magento2/blob/e6e6725329f75e36380e788e254ad012e8ae0759/app/code/Magento/Quote/Model/Quote/Item/AbstractItem.php#L178-L185

Instead, the `Validator` class should be checking the result of `getParentItem`

**Comment by aligent-lturner** [comment]
@magento give me 2.4-develop instance

**Comment by magento-engcom-team** [comment]
Hi @aligent-lturner. Thank you for your request. I'm working on Magento 2.4-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @aligent-lturner, here is your [Magento instance](https://i-26622-2-4-develop.instances.magento-community.engineering).
Admin access: [https://i-26622-2-4-develop.instances.magento-community.engineering/admin_9f98](https://i-26622-2-4-develop.instances.magento-community.engineering/admin_9f98)
Login: `277ecaa4` Password: `ef4621a07619`
Instance will be terminated in up to 3 hours.

**Comment by aligent-lturner** [comment]
Luma does not display discounts in minicart, and I do not have access to the underlying database to confirm the issue in the provided vanilla instance.

**Comment by aligent-lturner** [comment]
Actually - I was able to use the database backup functionality to confirm the issue in vanilla instance:

From quote table, after adding 2 $10 items:

```
INSERT INTO `quote` VALUES ('1',1,'2020-02-01 02:13:05','2020-02-01 02:17:19',NULL,1,0,0,'2','2.0000','0','0.0000','0.0000','USD','USD','USD','19.3300','19.3300',NULL,NULL,'3','0',NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,1,0,'180.150.113.242','1',NULL,NULL,NULL,'USD','1.0000','1.0000',NULL,NULL,'20.0000','20.0000','19.3300','19.3300','1',0,NULL,NULL,0);
```

Note the values $19.33 (i.e. a discount of $0.67 has been applied, rather than $1)

And from the quote_item table:
```
INSERT INTO `quote_item` VALUES ('3','1','2020-02-01 02:15:23','2020-02-01 02:17:19','3',1,NULL,0,'Test_Product-Blue','Test Product',NULL,'1',NULL,0,0,'1.0000','1.0000','10.0000','10.0000',NULL,'0.0000','0.3300','0.3300','0.0000','0.0000','0.0000','10.0000','10.0000','0.0000','1.0000','configurable',NULL,NULL,NULL,NULL,NULL,'10.0000','10.0000','10.0000','10.0000','0.0000','0.0000',NULL,0,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL),('4','1','2020-02-01 02:15:23','2020-02-01 02:15:23','1',1,'3',0,'Test_Product-Blue','Test Product-Blue',NULL,NULL,NULL,0,0,'1.0000','1.0000','10.0000','0.0000',NULL,'0.0000','0.0000','0.0000','0.0000','0.0000','0.0000','0.0000','0.0000','0.0000','0.0000','simple',NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,0,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL),('5','1','2020-02-01 02:17:19','2020-02-01 02:17:19','3',1,NULL,0,'Test_Product-Red','Test Product',NULL,'1',NULL,0,0,'1.0000','1.0000','10.0000','10.0000',NULL,'0.0000','0.3400','0.3400','0.0000','0.0000','0.0000','10.0000','10.0000','0.0000','1.0000','configurable',NULL,NULL,NULL,NULL,NULL,'10.0000','10.0000','10.0000','10.0000','0.0000','0.0000',NULL,0,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL),('6','1','2020-02-01 02:17:19','2020-02-01 02:17:19','2',1,'5',0,'Test_Product-Red','Test Product-Red',NULL,NULL,NULL,0,0,'1.0000','1.0000','10.0000','0.0000',NULL,'0.0000','0.0000','0.0000','0.0000','0.0000','0.0000','0.0000','0.0000','0.0000','0.0000','simple',NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,0,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL);
```

Here, we can see discounts for the 2 items of $0.33 and $0.34 instead of $0.50 for each.

**Comment by slavvka** [comment]
Hi @aligent-lturner. Thank you for your report. 
The issue has been fixed in magento/magento2#26623 by @aligent-lturner in 2.4-develop branch
Related commit(s):
 - [e943bda91d0b03a0fc67fb78b82254986796396c](https://github.com/magento/magento2/commit/e943bda91d0b03a0fc67fb78b82254986796396c)
 - [199afbff8651b267ba52a3536784e22cc37979f6](https://github.com/magento/magento2/commit/199afbff8651b267ba52a3536784e22cc37979f6)
 - [90c354044003d9d870f2ab52dbf0cf07ddde44b7](https://github.com/magento/magento2/commit/90c354044003d9d870f2ab52dbf0cf07ddde44b7)
 - [5f4cc5f6f81776db4419be934c1a3afd159fd38e](https://github.com/magento/magento2/commit/5f4cc5f6f81776db4419be934c1a3afd159fd38e)
 - [0e9c4ef9d7d541f8dba35052731f8d2d152fe4f7](https://github.com/magento/magento2/commit/0e9c4ef9d7d541f8dba35052731f8d2d152fe4f7)
 - [2c961afa3170257c4bed3c26d0d57d42d5445035](https://github.com/magento/magento2/commit/2c961afa3170257c4bed3c26d0d57d42d5445035)
 - [75b03e05c592075cf0619e419ac54d8bf62e34af](https://github.com/magento/magento2/commit/75b03e05c592075cf0619e419ac54d8bf62e34af)
 - [a037edfe8c14675a89e5e2f6839bf401ac9e253c](https://github.com/magento/magento2/commit/a037edfe8c14675a89e5e2f6839bf401ac9e253c)

The fix will be available with the upcoming 2.4.0 release.
